### DISCUSSION TITLE
Multiple TBT/USB4 retimer support - NDA

### DISCUSSION THREAD

**Discussion by divyassh** [fwupd/fwupd#3639] [Q&A]
**Description**

Multiple retimers are needed to support TBT/USB4 functionality for longer trace length between the Type-C connector and the SoC mux. This is a new change which will require firmware upgrade for multiple retimers on a single Type-C port. This case is only for No device attached.

**Solution:**

fwupd should be able to get the number of retimers and be able to update the firmware for each.
The retimer closest to the Type-C port should be upgraded first. The retimer firmware for the cascaded retimers needs to be assumed as separate binaries.

**Describe alternatives you've considered**

Currently this can be done manually. Steps in details below:
1. Boot up system with port 0 NDA (no device attached)
ls /sys/bus/thunderbolt/devices/0-0/usb4_port1, sysfs entries like "offline" "rescan" should be there.

2. Set up TBT connection
localhost /sys/bus/thunderbolt/devices/0-0/usb4_port1 # echo 1 > ./usb4_port1/offline

3. Scan retimers
localhost /sys/bus/thunderbolt/devices/0-0/usb4_port1 # echo 1 > ./usb4_port1/rescan
After scanning, you should have retimer entries 0-0:1.1 and 0-0:1.2.

localhost /sys/bus/thunderbolt/devices/0-0 # ls usb4_port1
0-0:1.1 0-0:1.2 firmware_node link offline power rescan uevent

4. Copy binary to connector side non-active nvmem and authenticate

localhost /sys/bus/thunderbolt/devices/0-0/usb4_port1 # dd if=<retimer_firmware_facing_connector.bin> of=./0-0:1.2/nvm_non_active1/nvmem

echo 2 > ./0-0:1.2/nvm_authenticate

Copy binary to soc side non-active nvmem and authenticate

localhost /sys/bus/thunderbolt/devices/0-0/usb4_port1 #dd if=<retimer_firmware_facing_soc.bin> of=./0-0:1.1/nvm_non_active0/nvmem

echo 2 > ./0-0:1.1/nvm_authenticate

5. Disconnect TBT connection
localhost /sys/bus/thunderbolt/devices/0-0/usb4_port1 # echo 0 > ./usb4_port1/offline

**Comment by hughsie** [comment]
This looks tricky to do in fwupd. We don't know what retimers are attached without a manual scan? Do the retimers give any vid/PID/guid we can use for firmware stream matching?

**Reply by divyassh** [reply]
For matching the firmware binary at this point it is standard suffix "facing_connector" and "facing_soc", which will work for two retimers.  Agreed that this doesnt scale well, needs to be thought about.

With current implementation, below sysfs will give the retimers connected after Step 2.
localhost /sys/bus/thunderbolt/devices/0-0 # ls usb4_port1
0-0:1.1 0-0:1.2

The numbering of retimer increases from SoC to connector, so "*facing_soc.bin" maps to "0-0:1.1" and "*facing_connector.bin" maps to 0-0:1.2.

**Reply by superm1** [reply]
Is it reasonable to perhaps encode the device sysfs filename in the GUID of those children devices? Then the mapping of binary go device is a metadata decision.

**Comment by superm1** [comment]
Is there not a way to toggle force power on these retimers to make them show up without a connection?

**Reply by divyassh** [reply]
So the step mentioned above will take care of the force power and is done without connecting the TBT device.

**Reply by superm1** [reply]
That was my suspicion, thanks for confirming.

**Comment by superm1** [comment]
Let me explain how fwupd works.

When the daemon starts up it will run `coldplug` routines that can enumerate any hardware that is not supported by a backend.  This can be things that don't unplug/replug/disappear.  The daemon will read these at startup and never again and they can't come and go.

Anything else will be event driven. This can be things like a USB device plugged in, a Udev entry showing from the kernel, or a BLE device coming within range.

Here's what I think you need during event detection:
1. Since fwupd already has support for Thunderbolt devices (and some retimer devices) I think what will be needed is when the event comes in for the parent device, if `offline` and `rescan` are present then it will need to call those to try to discover more "child" devices.  
2. Those child devices will need to have the parent set accordingly so that the furthest is a child to the closest
3. The policy will need to be set for parent to install updates first.
4. An extra flag (internal to the plugin - not a daemon flag) will need to be set on these to indicate that they need to be woken up by this _fake_ thunderbolt mode.

When it comes to flashing:
1. Check for this internal flag, if set then then for presence of a device connected or not
2. If no device connected, do this fake thunderbolt mode dance
3. Check for the sysfs file
4. Flash
5. Undo fake thunderbolt mode

That make sense?

**Reply by superm1** [reply]
One more thought; does the calling of offline and rescan `cause` uevents?  If so, perhaps just needing to run `rescan` when the sysfs file is available on an added parent device and ignoring the remove uevent would be sufficient.

**Reply by westeri** [reply]
It does yes. Writing 0 to offline will also remove all the connected retimers (Linux devices) so you get remove uevent from those. Running rescan also triggers add uevent for all the retimers found.

**Reply by superm1** [reply]
So I think then basically when the parent device uevent is processed (or at coldplug) if it sees the `rescan` attribute it will need to write 1 and all those child devices get added from uevent, but need to ignore the remove one from fwupd so it can keep it in device database what firmware was on the retimer even if not in fake tbt mode.

**Reply by djcampello** [reply]
cc @krankun

**Reply by krankun** [reply]
Sorry for being late I just came back from Spring break,

The current fwupd implementation for retimer NDA already supports steps 1 thru 5 , with some exceptions below 
1) step 4 the current implementation uses echo 1 for nvm_authenticate AFAIK , the need here is echo 2 > nvm_authenticate ?
2) for "The numbering of retimer increases from SoC to connector, so "*facing_soc.bin" maps to "0-0:1.1" and "*facing_connector.bin" maps to 0-0:1.2" 
[Kranthi] for this since we already have support for poking offline and rescan for usb4_port1 identifying this retimers with proper GUID's 
can help metadata in .cab files to make the decisions on the nvm flash,

I will start to get more hands-on once the hw is available and navigate the issue as I get more info ,

@divyassh is no longer working on ChromeOS , I left a note to Feng Li to join here
